ci/build_conda.sh

```diff
@@ -0,0 +1,19 @@
+#!/bin/bash
+# Copyright (c) 2024, NVIDIA CORPORATION
+
+set -euo pipefail
+
+source rapids-env-update
+
+export CMAKE_GENERATOR=Ninja
+
+rapids-print-env
+
+rapids-logger "Install CUDA Toolkit"
+source "$(dirname "$0")/install_latest_cuda_toolkit.sh"
```

Comment:
This process of installing system-wide cudatoolkit feels like an antipattern to me. It goes against what conda is good at, and I'm concerned that it might make things more brittle. As such, the work that we are needing to do now to address confusion with finding libraries feels like wasted effort.

The fundamental issue is that appropriate cudatoolkit libraries are not available on Conda, right? Is there a different approach we could take here, where we work on addressing that fundamental issue first and then use more standard conda patterns in this library?

Have I misunderstood what is going on here? Are there other reasons to be using a system install of the CTK?

Reply:
If we want to build with CUDA 12.2, those libraries are available

Reply:
The main issue is that the package needs to ship the latest version of `libnvjitlink_static` for it to be able to handle the PTX from the cuDF build. So far we've used the same pattern for the pip packages.

Reply:
I'm pretty confused here, so please correct me where I have no doubt gotten something wrong.

* We need pynvjitlink in order to finish our CUDA 12.2 release
* We can't use our existing parts of CUDA 12.2 to build pynvjitlink, because we need libnvjitlink from CUDA 12.3

Is that right? If so, I suppose you have no choice but the system packages, but it also seems like a really awkward position to be in. How was pynvjitlink both mandatory to 12.2 and not functional until 12.3?

Reply:
Great questions. I can try to illuminate the goals, design, and requirements, which might help.

nvjitlink is a library added in CUDA 12 that provides "a set of APIs which can be used at runtime to link together GPU device code". pynvjitlink provides Python bindings to that code. cuDF ships a bundle of PTX code generated at build time that is used by Numba and linked together with Numba-generated code. This process needs pynvjitlink in order to provide backward compatibility, if the PTX bundled with cuDF at build time is too new to run on the user's driver.

For example, if we build cuDF with CUDA 12.2, it will generate and bundle some CUDA 12.2 PTX code with the library. Numba wants to link this with user-provided code but the result will not run on driver 525 (which shipped alongside CUDA 12.0) because it doesn't know how to use the newer PTX code. In order for RAPIDS to migrate to building with CUDA 12.x (x > 0), we must depend on pynvjitlink.

The goal is to release a new pynvjitlink (which statically links libnvjitlink) for each new CUDA Toolkit minor version. (Other reasons to make a new version would include other matrix changes like Python version updates or updates to the pynvjitlink bindings themselves.) This release process is _not_ fundamentally synced to the RAPIDS release cycle, and will likely involve intermittent but off-cycle releases that are not aligned to RAPIDS 24.02, 24.04, etc.

We want (need?) to build with the newest possible version of nvjitlink for compatibility reasons. RAPIDS wants to move to building with CUDA 12.2, so it's reasonable to ask why we can't just use nvjitlink from CUDA 12.2 (which has conda-forge packages). There are some additional bugs in nvjitlink that were only recently resolved in 12.3 (see other thread: https://github.com/rapidsai/pynvjitlink/pull/33#discussion_r1445510368), so this isn't viable. We also want `pynvjitlink` to be usable by Numba consumers outside of RAPIDS, so supporting PTX from CUDA "12.latest" (and not just limiting to the CUDA build version used by RAPIDS) is important.

Generally, conda packages have lagged other channel sources (like RPM packages) in availability. Because we _need_ to build pynvjitlink against CUDA "12.latest" for pip users (who could have any driver/toolkit), we chose to pull nvjitlink from RPM packages. It is possible for conda packages of pynvjitlink to take a different strategy, but my initial hope was that we could just do the same thing and pull the latest CUDA RPM packages for the conda build. Clearly there are hiccups with CMake and conda support around this since it didn't work out of the box. I am certainly open to other strategies here but it will be difficult to maintain pynvjitlink if we are pulling the nvjitlink from two different sources with two different versions, because we want to make new releases for both pip and conda simultaneously.

We require the nvjitlink library from CUDA 12.3+ to proceed. CUDA 12.3 is not on `conda-forge` yet but that still wouldn't necessarily solve our problem if conda-forge continues to lag mainstream availability (we need to release pip and conda support together). Perhaps we could pull from `nvidia` channel packages or fix the issues with using system packages (which I trust to be released on "day one" of a new minor version).

Reply:
While discussing offline, @bdice and I realized the following case can occur. Suppose a user has:
- 12.0 Driver
- 12.2 cuDF
- 12.2u2 pynvjitlink
- 12.3 CTK
In this case even with the presence of `pynvjitlink` we will still encounter a failure since there's a piece of the PTX that the 12.3 CTK generates which will be too new for `libnvjitlink` from `12.2u2`. While conda provides some tools to constrain such that this would never happen, the same is not the case for pip packages. This provides some motivation to pursue the RPM package route.